analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Invocation.cs

```diff
@@ -40,10 +93,9 @@ protected override ProgramState[] Process(SymbolicContext context, IInvocationOp
         {
             state = state.ResetStaticFieldConstraints(invocation.TargetMethod.ContainingType);
         }
-        else if (invocation.Instance.TrackedSymbol(state) is { } symbol && !IsNullableGetValueOrDefault(invocation))
+        if (LearnNotNullCandidate(invocation) is { } operation && operation.TrackedSymbol(state) is { } symbol)
```

Comment:
During the Fixathon, we tried to rewrite this if like so:
```csharp
if (LearnNotNullCandidate(invocation) is { } operation)
{
    state = state.SetOperationConstraint(operation, ObjectConstraint.NotNull);
    if (operation.TrackedSymbol(state) is { } symbol)
    {
        state = state.SetSymbolConstraint(symbol, ObjectConstraint.NotNull);
    }
}
```
We had some tests failing, which looked weird and reverted. I reviewed these tests again and can confirm that this change causes a desired behavior! Please apply it again and update the UTs.

What is happening?
A `foreach` loop creates a hidden `try-catch` block. There, `collection.GetEnumerator` is being called and assigned to a flow capture. `.MoveNext()` and `.Current` are invoked on this flow capture. In the current implementation, both of these operations can throw and, therefore, branch into exception flows.
With the proposed change, the engine learns `NotNull` on the `FlowCaptureReference` operation when invoking `.MoveNext()`. Due to the logic in [ProgramState.SetOperationValue](https://github.com/SonarSource/sonar-dotnet/blob/6037c1ab81b8bc9d00d0c779147d14253e47302a/analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/ProgramState.cs#L86) the constraint is persisted on the underlying operation (the `.GetEnumerator()` invocation).
When `.Current` is analyzed, the `FlowCaptureOperation` is resolved to the same `InvocationOperation`, which now has the `NotNull` constraint. Therefore, the engine realizes that `.Current` will not throw and does not branch into an exception flow.
As a result, we avoid unnecessary branching, which is reflected in the failing UTs.